File: benchmarks/src/ruby/sort_by_comparator.rb
N=20000;arr=[];seed=12345;N.times{seed=(seed*16807)%2147483647;arr<<seed%1000000}
def qs(a,l,h,cmp)
 return if l>=h;s=[l,h]
 while s.any?
  h2=s.pop;l2=s.pop
  next if l2>=h2;p=a[(l2+h2)/2];i2,j2=l2,h2
  while i2<=j2
   i2+=1 while cmp.call(a[i2],p)<0;j2-=1 while cmp.call(a[j2],p)>0
   if i2<=j2;a[i2],a[j2]=a[j2],a[i2];i2+=1;j2-=1;end
  end
  if l2<j2; s.push(l2); s.push(j2); end
  if i2<h2; s.push(i2); s.push(h2); end
 end
end
asc=->a,b{a-b};desc=->a,b{b-a}
start=Time.now.to_f*1000
c1=arr.dup;qs(c1,0,N-1,asc);c2=arr.dup;qs(c2,0,N-1,desc)
r1=r2=0;N.times{|i|r1=(r1+c1[i]*(i%7))%1000003;r2=(r2+c2[i]*(i%7))%1000003}
finish=Time.now.to_f*1000
puts "Result: #{r1}_#{r2}";puts "Time: #{(finish-start).round}ms"

File: benchmarks/src/perl/sort_by_comparator.pl
use Time::HiRes qw(time);
my $N=20000;my @arr;my $seed=12345;for my $i (1..$N){$seed=($seed*16807)%2147483647;push @arr,$seed%1000000}
sub asc { $_[0] - $_[1] } sub desc { $_[1] - $_[0] }
sub qs { my($a,$l,$h,$cmp)=@_;return if $l>=$h;my @s=($l,$h);while(@s){my $h2=pop @s;my $l2=pop @s;next if $l2>=$h2;my $p=$a->[int(($l2+$h2)/2)];my($i2,$j2)=($l2,$h2);while($i2<=$j2){$i2++ while &$cmp($a->[$i2],$p)<0;$j2-- while &$cmp($a->[$j2],$p)>0;if($i2<=$j2){my $t=$a->[$i2];$a->[$i2]=$a->[$j2];$a->[$j2]=$t;$i2++;$j2--}}push @s,$l2,$j2 if $l2<$j2;push @s,$i2,$h2 if $i2<$h2}}
my $start=time();my @c1=@arr;qs(\@c1,0,$N-1,\&asc);my @c2=@arr;qs(\@c2,0,$N-1,\&desc);my($r1,$r2)=(0,0);
for my $i (0..$N-1){$r1=($r1+$c1[$i]*($i%7))%1000003;$r2=($r2+$c2[$i]*($i%7))%1000003}
my $end=time();my $ms=($end-$start)*1000;print "Result: ${r1}_${r2}\n";printf "Time: %.0fms\n",$ms;
